### PR TITLE
Number and string fixes

### LINKED ISSUE #9: Expose escape sequences as nodes
Hi, can you please have escape sequences (`\n`, `\t`, ...) be parsed into their own nodes so they can be highlighted? Some other parsers like [tree-sitter-c](https://github.com/tree-sitter/tree-sitter-c) already do this. You can find the list of valid escape sequences for Lua 5.4 under https://www.lua.org/manual/5.4/manual.html#3.1

### PR DESCRIPTION
**Pull request by amaanq**
Closes #9 

Should hopefully be much easier to go over, sorry about that :sweat_smile:

### CODE REVIEW

File: grammar.js [hunk 1]
```diff
@@ -321,6 +321,8 @@ module.exports = grammar({
       const hex_digits = /[a-fA-F0-9]+/;
       const hex_exponent_part = seq(choice('p', 'P'), signed_integer);
 
+      const bin_digits = /[01]+/;
```
Comment (MunifTanjim) [inline grammar.js:324]: Does Lua have the concept of binary digits? Is there any documentation?

File: grammar.js [hunk 2]
```diff
@@ -334,12 +336,13 @@ module.exports = grammar({
 
       const hex_literal = seq(
         choice('0x', '0X'),
+		    optional('.'),
```
Comment (MunifTanjim) [inline grammar.js:339]: I don't think lua accepts `.` after `0x` or `0X`.

File: grammar.js (hunk 1 above)
Comment (amaanq) [reply grammar.js:324]: I did not find any docs on it (yet), but here it seems to be valid: https://github.com/lua/lua/blob/e15f1f2bb7a38a3c94519294d031e48508d65006/testes/math.lua#L404

File: grammar.js (hunk 2 above)
Comment (MunifTanjim) [reply grammar.js:339]: Nevermind looks like Luajit supports it.

Comment (amaanq) [reply grammar.js:339]: https://github.com/lua/lua/blob/e15f1f2bb7a38a3c94519294d031e48508d65006/testes/math.lua#L507

Comment (MunifTanjim) [inline grammar.js:339]: Can you please add test for it? Just add an example here:

https://github.com/MunifTanjim/tree-sitter-lua/blob/0fc89962b7ff5c7d676b8592c1cbce1ceaa806fd/test/corpus/expressions.txt#L43-L60

File: grammar.js (hunk 1 above)
Comment (MunifTanjim) [reply grammar.js]: Yeah, should probably just update `decimal_integer_literal`.

File: grammar.js [hunk 3]
```diff
@@ -343,12 +344,58 @@ module.exports = grammar({
     },
 
     // LiteralString
-    string: ($) =>
+    string: $ => choice($._string_literal, $._multiline_string_literal),
```
Comment (MunifTanjim) [inline grammar.js:347]: This is a breaking change and I want to discuss it with https://github.com/nvim-treesitter/nvim-treesitter people first.

Can we please do it in a separate PR? That way the fixes for `number` won't be blocked by that discussion.

Comment (MunifTanjim) [reply grammar.js:347]: I've been putting of https://github.com/MunifTanjim/tree-sitter-lua/issues/9 all this time because I didn't want to break stuffs. Now that one of the core maintainers of `nvim-treesitter` is interested, it would be nice to make it happen! 😃

I'd still want to discuss it before making a sudden change, because it will also affect `neovim` core and other treesitter plugins. I don't want to break a lot of peoples editor without proper migration plan. Let's discuss it here and come up with a plan? https://github.com/nvim-treesitter/nvim-treesitter/discussions/4706

Comment (amaanq) [reply grammar.js:347]: Sure...but since we're not fundamentally changing any rules (only modifying the internals of string), just adding a capture for `@string.escape` (like I have in highlights.scm) would fix it.

File: grammar.js [hunk 4]
```diff
@@ -343,12 +344,58 @@ module.exports = grammar({
...
+
+    _string_literal: $ => choice(
       seq(
-        field('start', alias($._string_start, 'string_start')),
-        field('content', optional(alias($._string_content, 'string_content'))),
-        field('end', alias($._string_end, 'string_end'))
+        '"',
```
[the remark this answers is not in the document]
Comment (amaanq) [reply grammar.js:351]: Yes, with string content now being a named node, and the start/end bits being capturable anonymous nodes, there's no need for them (and imo they add bloat to the tree output..)

That does remind me that any queries capturing inner string contents will need to be updated to use the node not field now, I can take care of that in nvim-ts injections but I don't know (and really, I don't think) any plugins that would be affected by that

[the remark this answers is not in the document]
Comment (MunifTanjim) [reply grammar.js:351]: Are you still working on this PR? In case you're busy with something else, lemme know if you want me to take over.

[the remark this answers is not in the document]
Comment (amaanq) [reply grammar.js:351]: Yeah sorry just forgot to respond. I don't like this convention at all for several reasons:

The named nodes will consequentially just show a field named the same thing practically with an alias hiding the actual content with the name of the node..? That's really bizarre to me and is very unhelpful in debugging parsers, as well as serving no real purpose. `string_content` is already the same for both single and multiline strings via the alias, so that's a moot point, and modifying the queries should not be a factor to keep using worse practices

[the remark this answers is not in the document]
Comment (MunifTanjim) [reply grammar.js:351]: > The named nodes will consequentially just show a field named the same thing practically with an alias hiding the actual content with the name of the node..?

😅 I don't follow that argument. You're already willing to do that here: `alias($._multiline_string_start, '[[')`. `[[` won't always be the actual content, it can be `[=[`, `[==[` etc. If we're hiding the actual content already, what's the point?

Also, you don't really have to `alias` it to `string_start`, I'm only asking to keep the field name. you can just do `field('start', alias('"', '"'))` 


> string_content is already the same for both single and multiline strings via the alias, so that's a moot point

Sorry, but it's not.

For `local a = "a\0bc"`, the tree looks like this:

image

The same string has 2 `string_content` and each `escape_sequence` will keep adding more `string_content`. While block (multiline) string will always have only a single `string_content`.

[the remark this answers is not in the document]
Comment (amaanq) [reply grammar.js:351]: well, distinguishing [[ and [======[ is useless imo but I'd be fine to remove that alias for consistency if you'd like - having string content **is** useful however. 

The latter I can agree with having a single upper level node string_content which encapsulates unnamed nodes and escape sequences

[the remark this answers is not in the document]
Comment (MunifTanjim) [reply grammar.js]: Not sure what you did, but looks like the `escape_sequence` is completely gone from the tree now.

[the remark this answers is not in the document]
Comment (MunifTanjim) [reply grammar.js]: Anyway, thanks for your contribution! 😃

I think we have all the necessary things to make this work. I'll be making some tweaks to revert some unnecessary changes (e.g. many tests were changes without any good reason) and merge this as soon as possible.

[the remark this answers is not in the document]
Comment (amaanq) [reply grammar.js]: > Not sure what you did, but looks like the `escape_sequence` is completely gone from the tree now.

No? See:
image

[the remark this answers is not in the document]
Comment (amaanq) [reply grammar.js]: > (e.g. many tests were changes without any good reason)

Well, the tests would fail otherwise with the grammar changes :laughing: 

As for renaming stuff, I'd like the named node names to not changed, the unnamed (hidden) ones, feel free to edit

[the remark this answers is not in the document]
Comment (MunifTanjim) [reply grammar.js]: > No? See:

Did you install with the latest changes? 🤔

In you screenshot, tehre are named `escape_sequence` node,

image

but these tests are passing and there are no `escape_sequence` node inside the `string_content` node.

[the remark this answers is not in the document]
Comment (clason) [reply grammar.js]: (@amaanq I've seen that on another PR, too: the changes to the generated files you push are not the ones I get if I run `tree-sitter generate` locally. Might want to investigate that.)

File: grammar.js [hunk 5]
```diff
@@ -343,12 +344,52 @@ module.exports = grammar({
...
     // LiteralString
-    string: ($) =>
-      seq(
-        field('start', alias($._string_start, 'string_start')),
-        field('content', optional(alias($._string_content, 'string_content'))),
-        field('end', alias($._string_end, 'string_end'))
-      ),
+    string: $ => choice($._string_literal, $._multiline_string_literal),
+
+    _string_literal: $ => choice(
```
Comment (MunifTanjim) [inline grammar.js:349]: Found another issue:

```lua
local a = "a
b"
``` 

This is now valid. Althrough this is not a valid lua string.

File: grammar.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (MunifTanjim) [reply grammar.js]: I've resolved this. But found another issue: https://github.com/MunifTanjim/tree-sitter-lua/pull/20#discussion_r1181493390

**Comment by clason** [comment]
@MunifTanjim I've been meaning to ask (and I apologize for the tangent): Is there a concrete reason you're still sticking to ABI 13 for this parser?

**Comment by MunifTanjim** [comment]
> @MunifTanjim I've been meaning to ask (and I apologize for the tangent): Is there a concrete reason you're still sticking to ABI 13 for this parser?

Nope. Happy to upgrade. Just that nobody asked for it yet 😂

**Comment by clason** [comment]
Then I do 🤗

File: grammar.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (amaanq) [reply grammar.js]: probably because I use ahlincs ts fork, but it shouldn't have any major differences at all - \z isn't a valid escape sequence FYI

[the remark this answers is not in the document]
Comment (MunifTanjim) [reply grammar.js]: > it shouldn't have any major differences at all - \z isn't a valid escape sequence FYI

It is tho. You can check the reference manual: https://www.lua.org/manual/5.4/manual.html#3.1

> The escape sequence '\z' skips the following span of whitespace characters, including line breaks; it is particularly useful to break and indent a long literal string into multiple lines without adding the newlines and spaces into the string contents.

[the remark this answers is not in the document]
Comment (clason) [reply grammar.js]: It's not in Lua 5.1, tho. Maybe the language version this parser targets should be explicitly noted in the README?

[the remark this answers is not in the document]
Comment (MunifTanjim) [reply grammar.js]: Hmm... So far, I've only tested it against 5.x syntax.

[the remark this answers is not in the document]
Comment (clason) [reply grammar.js]: > 5.x

That's not a thing, though -- 5.1, 5.2, 5.3, and 5.4 are basically different language dialects (more like Python 2 vs. Python 3).

[the remark this answers is not in the document]
Comment (MunifTanjim) [reply grammar.js]: As far as I can say, the syntax is mostly backward compatible? as in, syntax for 5.1 is valid in 5.4?

Not sure how tree-sitter parsers for language like lua should be. Should there be 4 different parser for `5.1` - `5.4`? 🤔

[the remark this answers is not in the document]
Comment (MunifTanjim) [reply grammar.js]: There's also LuaJIT. 😂 `\z` is valid in LuaJIT... that's how I discovered it, when working with codes supposed to run with LuaJIT.

[the remark this answers is not in the document]
Comment (clason) [reply grammar.js]: and `luau`...

I don't know what the best approach here is; supporting Lua 5.4 hasn't been a problem for us so far.
But I do think documentation is never wrong.

[the remark this answers is not in the document]
Comment (MunifTanjim) [reply grammar.js]: well, `luau` has some major changes. Like `if` can be used as expression.

> But I do think documentation is never wrong.

Yep, will add that in README.

File: grammar.js (hunk 5 above)
Comment (amaanq) [reply grammar.js]: Gotcha, thanks! Will fix soon

File: grammar.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (amaanq) [reply grammar.js]: Luau I can extend and modify it accordingly, but I don't think minute false positives matter so much like Clason said prior

File: grammar.js (hunk 5 above)
Comment (MunifTanjim) [reply grammar.js]: Are these valid lua escaped string?

`/u[0-9a-fA-F]{4}/`, `/U[0-9a-fA-F]{8}/` 

Something like this should be more appropriate? 🤔

```js
    escape_sequence: () =>
      token(
        prec(
          1,
          seq(
            '\\',
            choice(
              /[\nabfnrtv\\'"]/,
              /z\s*/,
              /[0-9]{1,3}/,
              /x[0-9a-fA-F]{2}/,
              /u\{[0-9a-fA-F]+\}/
            )
          )
        )
      ),
```

Comment (amaanq) [reply grammar.js:349]: Sorry for the delay last weeks of school was hectic, I've fixed \z and \090, bizarre cases in general

**Comment by amaanq** [comment]
Do check it out now, should have all the bases covered

File: grammar.js [hunk 6]
```diff
@@ -343,12 +344,52 @@ module.exports = grammar({
...
+      prec(1, $.escape_sequence),
+    ),
+
+    escape_sequence: _ => token.immediate(seq(
+      '\\',
+      choice(
+        /[^xu0-7]/,
+        /[0-9]{1,3}/,
+        /x[0-9a-fA-F]{2}/,
+        /u[0-9a-fA-F]{4}/,
```
Comment (MunifTanjim) [inline grammar.js:388]: Just double checking... are these valid lua escaped string?

`/u[0-9a-fA-F]{4}/`, `/U[0-9a-fA-F]{8}/`

Comment (amaanq) [reply grammar.js:388]: I believe 5.3 added support for it

Comment (MunifTanjim) [reply grammar.js:388]: I couldn't get them to work with any of Lua 5.x or LuaJIT.

Comment (amaanq) [reply grammar.js:388]: My mistake, it requires brackets, so `\u{0090}`

**Comment by clason** [comment]
(Once you make a release, I'll bump the parser in Neovim and sync the queries from nvim-treesitter; that'll pull in your previous fixes as well.)

(On that note, can I interest you in a "semver for parsers"? ;) 

* point releases for parser changes that do not affect queries at all (grammar.js or scanner.c bugfixes),
* minor releases which add nodes (meaning old queries will still work, but need changes to make full use of the updated grammar),
* major releases which remove or rename nodes (meaning old queries will no longer work).

Would make my life easier so I could blindly merge some updates...)

**Comment by MunifTanjim** [comment]
> On that note, can I interest you in a "semver for parsers"?

Absolutely. Let me just add a few more lua codebases for the tests to parse. Once I'm a little more sure that the whole 5.x syntax is covered, I'll cut a `1.x` release.

I think these will be the next few releases:
- the current grammar with ABI 13 (this weekend)
- the same grammar with ABI 14 (immediately after the previous one)
- add more testing and do 1.x (might be a few weeks)

**Comment by clason** [comment]
No hurries, just a suggestion -- and one that can be followed without requiring a 1.0 release (treever need not be semver).
